i2c: Bug fixes for default RTIO handler

teburd opened the pull request:
TXRX is meant specifically to handle a full duplex bus like SPI, I2C is half duplex meaning only read or write can be performed at once. Drop TXRX from the default handler.

Each transaction should result in a single call to i2c_transfer. This corrects the default handler to do just that. Also simplifies the various helpers here.

Commit:
i2c: Drop TXRX from default RTIO handler

TXRX is meant specifically to handle a full duplex bus like SPI, I2C is
half duplex meaning only read or write can be performed at once.

Drop TXRX as a supported operation code for the default I2C submission
path.

Signed-off-by: Tom Burdick <thomas.burdick@intel.com>

ubieda commented:
Not sure why this PR didn't get review-requests.

Commit:
i2c: Fix default RTIO handler transactions

Transactions from RTIO should result in single calls to i2c_transfer.
This corrects the default handler to first count the number of
submissions in the transaction, allocate on the stack, and then copy
over each submission to an equivalent i2c_msg.

It also cleans up the helper functions to be infallible, taking only the
submission and msg to copy to.

Signed-off-by: Tom Burdick <thomas.burdick@intel.com>

XenuIsWatching commented:
looks like some tests may need to be updated... but I ported over this code to my default i3c rtio handler branch that I'm working on and it's looking as expected with 'restarts' in between transactions. Thanks for this!!

ubieda reviewed (dismissed):
Code looks good! Approving assuming we take care of https://github.com/zephyrproject-rtos/zephyr/tree/main/tests/subsys/rtio/rtio_i2c.

bjarki-andreasen commented on drivers/i2c/i2c_rtio_default.c:
```diff
@@ -12,97 +12,115 @@
+	}
+
+	/* Allocate msgs on the stack */
+	struct i2c_msg msgs[num_msgs];
```
Variable length array types are not MISRA compliant :) not sure why CI is not complaining :) I ran into issues with it here https://github.com/zephyrproject-rtos/zephyr/pull/77552

teburd replied:
I'm open to suggestions here... I didn't want to bother with heap. I guess a statically sized array on the stack? painful.

ubieda replied:
How about a Memslab Pool: You'd only take it while performing the task, so hopefully is low-ish risk?

teburd replied:
Made it a Kconfig as suggested by @bjarki-andreasen with a default of 4

bjarki-andreasen commented on drivers/i2c/i2c_rtio_default.c:
```diff
@@ -12,97 +12,126 @@
+	* in a transaction as we typically would only have 2, one to write a
+	* register address, and another to read/write the register into an array
+	*/
+	if (num_msgs > 4) {
```
maybe overlay pedantic, but maybe define the 4? `#define RTIO_I2C_MSG_ALLOC 4` and even better, make it a Kconfig?

teburd replied:
yes, done

Commit:
i2c: Drop transceive test, fix transfer call count

Transactions should result in a single transfer call not multiple
transfer calls. Transceive isn't supported by i2c and so the TXRX op
isn't validated for success anymore.

Signed-off-by: Tom Burdick <thomas.burdick@intel.com>

bjarki-andreasen reviewed (approved):
Nice!
